### PR TITLE
add custom replacer

### PR DESCRIPTION
**Pull request by lilee**
as mentioned in the issue #408

### CODE REVIEW

File: middleware/replacer.go [hunk 1]
```diff
@@ -117,3 +123,14 @@ const (
...
+
+var (
+	customReplacements      = map[string]string{}
+	customReplacementsMutex sync.RWMutex
+)
+
+func RegisterReplacement(name, value string) {
+	customReplacementsMutex.Lock()
+	defer customReplacementsMutex.Unlock()
+	customReplacements[name] = value
```
Comment (mholt) [inline middleware/replacer.go]: So is there really a race condition without the mutex? Does your add-on call RegisterReplacement during request-time or only at the initial setup phase? (Mostly interested in the necessity of the mutex for performance reasons.)

Comment (lilee) [inline middleware/replacer.go]: I call it in request-time, so it will produce a race without mutex.

**Comment by mholt** [comment]
After looking a little more at this and chatting with @abiosoft, a few thoughts. I think this would be better if it were instance-scoped instead of package-scoped. What I mean is that instead of a "global" registry of replacers -- which could greatly increase lock contention and conflict with various hosts using the same facility -- you make each new replacer with its custom values; something like:

```go
replacer := middleware.NewReplacer(…)
replacer.Set(key, value)
```

If you want something global, environment variables are already there for that.

Also, I'm not sure that the "@" syntax is necessary.

**Comment by abiosoft** [comment]
There could be an easy way out.

If setting of value and proxy is happening at same request, custom headers is the way to go. 

A good example is **rewrite** setting `Caddy-Rewrite-Original-URI` that is later consumed by **fastcgi**.

**Comment by mholt** [comment]
Something like that is good, too. I actually really like that -- just have your custom middleware set a header on the request that proxy then proxies through. (Won't require any changes to replacer or proxy middleware)

**Comment by mholt** [comment]
@lilee What do you think about just setting a header from your middleware instead?

**Comment by mholt** [comment]
Closing due to inactivity, but feel free to discuss if there's anything else.

**Pull request closed without merging**